### PR TITLE
Parse task queue stats from /proc/schedstat

### PR DESCRIPTION
**Pull request by bitglue**
These are useful as an indication of CPU contention and scheduler
latency.

### CODE REVIEW

**Comment by bitglue** [comment]
@discordianfish @pgier ready for review.

I'll update https://github.com/prometheus/node_exporter/pull/1389 to use this after review.

File: schedstat.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+type Schedstat struct {
+	CPUs []*SchedstatCPU
+}
+
+// SchedstatCPU contains the values from one "cpu<N>" line
+type SchedstatCPU struct {
+	CPUNum string
+
+	RunningJiffies uint64
+	RunningSeconds float64
```
Comment (pgier) [inline schedstat.go]: What do you think about adding a method to dynamically provide the seconds instead of storing both jiffies and seconds in the struct?

```golang
func (s SchedstatCPU) RunningSeconds() float64 {
	return float64(s.RunningJiffies) / userHZ
}
```

File: schedstat.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+func (fs FS) Schedstat() (*Schedstat, error) {
+	file, err := os.Open(fs.proc.Path("schedstat"))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	stats := &Schedstat{}
+	scanner := bufio.NewScanner(file)
+
```
Comment (pgier) [inline schedstat.go]: Would it be helpful to check the version number at the top of the file, or is the format consistent enough between versions that it's not necessary?

Comment (bitglue) [reply schedstat.go]: I think there would be little practical value in doing so. Version 15 is current, and according to the documentation:

> Version 15 of schedstats dropped counters for some sched_yield:
yld_exp_empty, yld_act_empty and yld_both_empty. Otherwise, it is
identical to version 14.

Version 14 seems to cover kernels at least 10 years back, and is compatible. So to support any older versions would be a bit of an archaeology project.

Newer versions may not remain compatible, but probably more likely than not they will be compatible enough for the subset of metrics extracted here.

Comment (pgier) [reply schedstat.go]: Ok, sounds good.

**Comment by bitglue** [comment]
@pgier Everything addressed I think, please re-review

**Comment by bitglue** [comment]
@discordianfish @pgier bump, review please

File: schedstat.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		}
+	}
+
+	return stats, nil
+}
+
+func parseProcSchedstat(filename string) (ProcSchedstat, error) {
+	stats := ProcSchedstat{}
+
+	file, err := os.Open(filename)
```
Comment (pgier) [inline schedstat.go]: I think it would be better to use `ioutil.ReadFile`, then you don't need the loop and scanner.

Comment (bitglue) [reply schedstat.go]: good idea, one minute...

**Comment by bitglue** [comment]
@pgier ok, sorry to change things just after asking for review. It occured to me reading the diff with fresh eyes and a cup of coffee that there were untested corner cases. I've added some more tests and rearranged some things. Should be good now, for reals.

**Comment by pgier** [comment]
Merged, thanks!

**Comment by bitglue** [comment]
What's the process for getting a released tagged so I can complete https://github.com/prometheus/node_exporter/pull/1389 ?

**Comment by bitglue** [comment]
Actually now that I've learned a bit more about go modules I've figured out how to update it without a release. Seems I'm not the first one to pin node_exporter to a non-released version of procfs.
